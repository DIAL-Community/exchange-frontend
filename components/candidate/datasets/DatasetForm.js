import React, { useCallback, useContext, useEffect, useMemo, useRef, useState } from 'react'
import { useRouter } from 'next/router'
import { useMutation } from '@apollo/client'
import { useIntl } from 'react-intl'
import { FaSpinner } from 'react-icons/fa'
import { Controller, useForm } from 'react-hook-form'
import { validate } from 'email-validator'
import ReCAPTCHA from 'react-google-recaptcha'
import Link from 'next/link'
import { HtmlEditor } from '../../shared/HtmlEditor'
import Input from '../../shared/Input'
import { ToastContext } from '../../../lib/ToastContext'
import ValidationError from '../../shared/ValidationError'
import Select from '../../shared/Select'
import { useUser } from '../../../lib/hooks'
import { CREATE_CANDIDATE_DATASET } from '../../../mutations/dataset'
import UrlInput from '../../shared/UrlInput'
import { Unauthorized } from '../../shared/FetchStatus'
import { getDatasetTypeOptions } from '../../../lib/utilities'
import { BREADCRUMB_SEPARATOR } from '../../shared/breadcrumb'

const DatasetForm = () => {
  const { formatMessage } = useIntl()
  const format = useCallback((id, values) => formatMessage({ id }, values), [formatMessage])

  const router = useRouter()

  const { user } = useUser()

  const [mutating, setMutating] = useState(false)
  const [reverting, setReverting] = useState(false)

  const captchaRef = useRef(null)

  const { showToast } = useContext(ToastContext)
  const { locale } = useRouter()

  const [createCandidateDataset, { reset }] = useMutation(CREATE_CANDIDATE_DATASET, {
    onError: () => {
      setMutating(false)
      showToast(
        format('ui.candidateDataset.submit.failure'),
        'error',
        'top-center',
        1000
      )
      reset()
    },
    onCompleted: () => {
      setMutating(false)
      showToast(
        format('ui.candidateDataset.submit.success'),
        'success',
        'top-center',
        1000,
        null,
        () => router.push('/datasets')
      )
    }
  })

  const datasetTypeOptions = useMemo(() => getDatasetTypeOptions(format), [format])

  const { handleSubmit, register, control, setValue, formState: { errors } } = useForm({
    mode: 'onSubmit',
    reValidateMode: 'onChange',
    shouldUnregister: true,
    defaultValues: {
      name: '',
      website: '',
      visualizationUrl: '',
      datasetType: datasetTypeOptions[0],
      submitterEmail: '',
      description: '',
      captcha: null
    }
  })

  const doUpsert = async (data) => {
    if (user) {
      setMutating(true)

      const captcha = captchaRef.current.getValue()

      const { userEmail, userToken } = user
      const {
        name,
        website,
        visualizationUrl,
        datasetType,
        submitterEmail,
        description
      } = data

      createCandidateDataset({
        variables: {
          name,
          slug: '',
          website,
          visualizationUrl,
          datasetType: datasetType.value,
          submitterEmail,
          description,
          captcha
        },
        context: {
          headers: {
            'Accept-Language': locale,
            Authorization: `${userEmail} ${userToken}`
          }
        }
      })
    }
  }

  const cancelForm = () => {
    setReverting(true)
    router.push('/datasets')
  }

  useEffect(() => {
    register('captcha', { required: format('validation.required') })
  }, [register, format])

  return user ? (
    <div className='flex flex-col'>
      <div className='hidden lg:block px-8'>
        <div className='bg-white pb-3 lg:py-4 whitespace-nowrap text-ellipsis overflow-hidden'>
          <Link href='/' className='inline text-dial-blue h5'>
            {format('app.home')}
          </Link>
          <div className='inline h5'>
            {BREADCRUMB_SEPARATOR}
            <Link href='/datasets' className='text-dial-blue'>
              {format('ui.dataset.header')}
            </Link>
            {BREADCRUMB_SEPARATOR}
            <span className='text-dial-gray-dark'>
              {format('app.create')}
            </span>
          </div>
        </div>
      </div>
      <div className='pb-8 px-8'>
        <div id='content' className='sm:px-0 max-w-full mx-auto'>
          <form onSubmit={handleSubmit(doUpsert)}>
            <div className='bg-edit shadow-md rounded px-8 pt-6 pb-12 mb-4 flex flex-col gap-3'>
              <div className='text-2xl font-semibold text-dial-sapphire pb-4'>
                {format('ui.candidateDataset.label')}
              </div>
              <div className='flex flex-col lg:flex-row gap-4'>
                <div className='w-full lg:w-1/2 flex flex-col gap-y-3'>
                  <div className='form-field-wrapper' data-testid='candidate-dataset-name'>
                    <label className='form-field-label required-field' htmlFor='name'>
                      {format('ui.dataset.name')}
                    </label>
                    <Input
                      {...register('name', { required: format('validation.required') })}
                      id='name'
                      placeholder={format('ui.dataset.name')}
                      isInvalid={errors.name}
                      data-testid='dataset-name-input'
                    />
                    {errors.name && <ValidationError value={errors.name?.message} />}
                  </div>
                  <div className='form-field-wrapper' data-testid='candidate-dataset-website'>
                    <label className='form-field-label required-field' htmlFor='website'>
                      {format('ui.dataset.website')}
                    </label>
                    <Controller
                      name='website'
                      control={control}
                      render={({ field: { value, onChange } }) => (
                        <UrlInput
                          value={value}
                          onChange={onChange}
                          id='website'
                          isInvalid={errors.website}
                          placeholder={format('ui.dataset.website')}
                        />
                      )}
                      rules={{ required: format('validation.required') }}
                    />
                    {errors.website && <ValidationError value={errors.website?.message} />}
                  </div>
                  <div className='form-field-wrapper'>
                    <label className='form-field-label' htmlFor='dataVisualizationUrl'>
                      {format('ui.dataset.visualizationUrl')}
                    </label>
                    <Controller
                      name='dataVisualizationUrl'
                      control={control}
                      render={({ field: { value, onChange } }) => (
                        <UrlInput
                          value={value}
                          onChange={onChange}
                          id='dataVisualizationUrl'
                          placeholder={format('ui.dataset.visualizationUrl')}
                        />
                      )}
                    />
                  </div>
                  <div className='form-field-wrapper'>
                    <label className='form-field-label'>
                      {format('ui.dataset.datasetType')}
                    </label>
                    <Controller
                      name='dataType'
                      control={control}
                      render={({ field }) =>
                        <Select {...field} options={datasetTypeOptions} placeholder={format('ui.dataset.datasetType')} />
                      }
                    />
                  </div>
                  <div className='form-field-wrapper' data-testid='candidate-dataset-email'>
                    <label className='form-field-label required-field' htmlFor='email'>
                      {format('app.email')}
                    </label>
                    <Input
                      {...register('submitterEmail',
                        {
                          required: format('validation.required'),
                          validate: value => validate(value) || format('validation.email')
                        }
                      )}
                      id='email'
                      placeholder={format('app.email')}
                      isInvalid={errors.submitterEmail}
                    />
                    {errors.submitterEmail && <ValidationError value={errors.submitterEmail?.message} />}
                  </div>
                  <div className='form-field-wrapper'>
                    <ReCAPTCHA
                      sitekey={process.env.NEXT_PUBLIC_CAPTCHA_SITE_KEY}
                      ref={captchaRef}
                      onChange={value => {
                        setValue('captcha', value, { shouldValidate: true })
                      }}
                    />
                    {errors.captcha && <ValidationError value={errors.captcha?.message} />}
                  </div>
                </div>
                <div className='w-full lg:w-1/2'>
                  <div className='form-field-wrapper' data-testid='candidate-dataset-description'>
                    <label className='form-field-label required-field'>
                      {format('ui.dataset.description')}
                    </label>
                    <Controller
                      name='description'
                      control={control}
                      render={({ field: { value, onChange } }) => (
                        <HtmlEditor
                          editorId='description-editor'
                          onChange={onChange}
                          initialContent={value}
                          placeholder={format('ui.dataset.description')}
                          isInvalid={errors.description}
                        />
                      )}
                      rules={{ required: format('validation.required') }}
                    />
                    {errors.description && <ValidationError value={errors.description?.message} />}
                  </div>
                </div>
              </div>
              <div className='flex flex-wrap text-xl mt-8 gap-3'>
                <button
                  type='submit'
                  className='submit-button'
                  disabled={mutating || reverting}
                  data-testid='submit-button'
                >
                  {format('ui.candidateDataset.submit')}
                  {mutating && <FaSpinner className='spinner ml-3' />}
                </button>
                <button
                  type='button'
                  className='cancel-button'
                  disabled={mutating || reverting}
                  onClick={cancelForm}
                >
                  {format('app.cancel')}
                  {reverting && <FaSpinner className='spinner ml-3' />}
                </button>
              </div>
            </div>
          </form>
        </div>
      </div>
    </div>
  ) : <Unauthorized />
}

export default DatasetForm
