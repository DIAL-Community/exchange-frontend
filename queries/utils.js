const MIN_SEARCH_VALUE_CHARACTERS = 2

export const fetchSelectOptions = async (client, input, query, fetchedDataCallback, locale = null) => {
  const variables = {
    search: input
  }

  if (input && input.trim().length < MIN_SEARCH_VALUE_CHARACTERS) {
    return []
  } else if (locale) {
    variables.locale = locale
  }

  const response = await client.query({ query, variables })

  return response.data ? fetchedDataCallback(response.data) : []
}

export const fetchSelectOptionsWithMature = async (client, input, query, fetchedDataCallback, locale = null) => {
  const variables = {
    search: input,
    mature: true
  }

  if (input && input.trim().length < MIN_SEARCH_VALUE_CHARACTERS) {
    return []
  } else if (locale) {
    variables.locale = locale
  }

  const response = await client.query({ query, variables })

  return response.data ? fetchedDataCallback(response.data) : []
}
