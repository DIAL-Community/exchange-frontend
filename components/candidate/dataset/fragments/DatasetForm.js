import { memo, useCallback, useContext, useMemo, useState } from 'react'
import { useRouter } from 'next/router'
import { Controller, useForm } from 'react-hook-form'
import { FaSpinner } from 'react-icons/fa6'
import { useIntl } from 'react-intl'
import { useMutation } from '@apollo/client'
import { GRAPH_QUERY_CONTEXT } from '../../../../lib/apolloClient'
import { ToastContext } from '../../../../lib/ToastContext'
import { CustomMCaptcha } from '../../../shared/CustomMCaptcha'
import { HtmlEditor } from '../../../shared/form/HtmlEditor'
import Input from '../../../shared/form/Input'
import { generateDatasetTypeOptions } from '../../../shared/form/options'
import Select from '../../../shared/form/Select'
import UrlInput from '../../../shared/form/UrlInput'
import ValidationError from '../../../shared/form/ValidationError'
import { CREATE_CANDIDATE_DATASET } from '../../../shared/mutation/candidateDataset'
import {
  CANDIDATE_DATASET_PAGINATION_ATTRIBUTES_QUERY, PAGINATED_CANDIDATE_DATASETS_QUERY
} from '../../../shared/query/candidateDataset'
import { DEFAULT_PAGE_SIZE } from '../../../utils/constants'

const DatasetForm = memo(({ dataset }) => {
  const { formatMessage } = useIntl()
  const format = useCallback((id, values) => formatMessage({ id }, values), [formatMessage])

  const slug = dataset?.slug ?? ''

  const [mutating, setMutating] = useState(false)
  const [reverting, setReverting] = useState(false)
  const [captchaToken, setCaptchaToken] = useState()

  const { showSuccessMessage, showFailureMessage } = useContext(ToastContext)

  const router = useRouter()
  const { locale } = router

  const [updateDataset, { reset }] = useMutation(CREATE_CANDIDATE_DATASET, {
    refetchQueries: [{
      query: CANDIDATE_DATASET_PAGINATION_ATTRIBUTES_QUERY,
      variables: { search: '' },
      context: {
        headers: {
          ...GRAPH_QUERY_CONTEXT.VIEWING
        }
      }
    }, {
      query: PAGINATED_CANDIDATE_DATASETS_QUERY,
      variables: { search: '', limit: DEFAULT_PAGE_SIZE, offset: 0 },
      context: {
        headers: {
          ...GRAPH_QUERY_CONTEXT.VIEWING
        }
      }
    }],
    onCompleted: (data) => {
      const { createCandidateDataset: response } = data
      if (response.candidateDataset && response.errors.length === 0) {
        setMutating(false)
        const redirectPath = `/${locale}` +
          `/candidate/datasets/${response.candidateDataset.slug}`
        const redirectHandler = () => router.push(redirectPath)
        showSuccessMessage(
          format('toast.submit.success', { entity: format('ui.candidateDataset.label') }),
          redirectHandler
        )
      } else {
        showFailureMessage(format('toast.submit.failure', { entity: format('ui.candidateDataset.label') }))
        setMutating(false)
        reset()
      }
    },
    onError: () => {
      showFailureMessage(format('toast.submit.failure', { entity: format('ui.candidateDataset.label') }))
      setMutating(false)
      reset()
    }
  })

  const datasetTypeOptions = useMemo(() => generateDatasetTypeOptions(format), [format])

  const {
    handleSubmit,
    register,
    control,
    formState: { errors }
  } = useForm({
    mode: 'onBlur',
    reValidateMode: 'onChange',
    shouldUnregister: true,
    defaultValues: {
      name: dataset?.name ?? '',
      website: dataset?.website ?? '',
      visualizationUrl: dataset?.visualizationUrl ?? '',
      datasetType: datasetTypeOptions.find(({ value }) => value === dataset?.datasetType) ?? datasetTypeOptions[0],
      submitterEmail: dataset?.submitterEmail ?? '',
      description: dataset?.description ?? ''
    }
  })

  const doUpsert = async (data) => {
    // Set the loading indicator.
    setMutating(true)
    // Pull all needed data from session and form.
    const {
      name,
      website,
      visualizationUrl,
      datasetType,
      submitterEmail,
      description
    } = data
    // Send graph query to the backend. Set the base variables needed to perform update.
    const variables = {
      name,
      slug,
      website,
      visualizationUrl,
      datasetType: datasetType.value,
      submitterEmail,
      description,
      captcha: captchaToken
    }

    updateDataset({
      variables,
      context: {
        headers: {
          'Accept-Language': locale
        }
      }
    })
  }

  const cancelForm = () => {
    if (dataset) {
      setReverting(true)
      router.push(`/${locale}/candidate/datasets/${slug}`)
    }
  }

  return (
    <form onSubmit={handleSubmit(doUpsert)}>
      <div className='px-4 lg:px-0 py-4 lg:py-6 text-dial-meadow'>
        <div className='flex flex-col gap-y-6 text-sm'>
          <div className='text-xl font-semibold'>
            {dataset
              ? format('app.editEntity', { entity: dataset.name })
              : `${format('app.createNew')} ${format('ui.dataset.label')}`}
          </div>
          <div className='flex flex-col gap-y-2'>
            <label className='required-field' htmlFor='name'>
              {format('ui.dataset.name')}
            </label>
            <Input
              {...register('name', { required: format('validation.required') })}
              id='name'
              placeholder={format('ui.dataset.name')}
              isInvalid={errors.name}
            />
            {errors.name && <ValidationError value={errors.name?.message} />}
          </div>
          <div className='flex flex-col gap-y-2'>
            <label className='required-field' htmlFor='website'>
              {format('ui.candidateDataset.website.hint')}
            </label>
            <Controller
              name='website'
              control={control}
              render={({ field: { value, onChange } }) => (
                <UrlInput
                  value={value}
                  onChange={onChange}
                  id='website'
                  isInvalid={errors.website}
                  placeholder={format('ui.dataset.website')}
                />
              )}
              rules={{ required: format('validation.required') }}
            />
            {errors.website && <ValidationError value={errors.website?.message} />}
          </div>
          <div className='flex flex-col gap-y-2'>
            <label htmlFor='visualizationUrl'>
              {format('ui.candidateDataset.visualizationUrl.hint')}
            </label>
            <Controller
              name='visualizationUrl'
              control={control}
              render={({ field: { value, onChange } }) => (
                <UrlInput
                  value={value}
                  onChange={onChange}
                  id='visualizationUrl'
                  placeholder={format('ui.dataset.visualizationUrl')}
                />
              )}
            />
          </div>
          <div className='flex flex-col gap-y-2'>
            <label htmlFor='react-select-dataset-type'>
              {format('ui.candidateDataset.datasetType.hint')}
            </label>
            <Controller
              name='datasetType'
              control={control}
              render={({ field }) =>
                <Select {...field}
                  name='dataset-type'
                  options={datasetTypeOptions}
                  placeholder={format('ui.dataset.datasetType')}
                />
              }
            />
          </div>
          <div className='flex flex-col gap-y-2'>
            <label id='description' className='required-field'>
              {format('ui.dataset.description')}
            </label>
            <Controller
              name='description'
              control={control}
              render={({ field: { value, onChange } }) => (
                <HtmlEditor
                  labelledBy='description'
                  onChange={onChange}
                  initialContent={value}
                  placeholder={format('ui.dataset.description')}
                />
              )}
              rules={{ required: format('validation.required') }}
            />
            {errors.description && <ValidationError value={errors.description?.message} />}
          </div>
          <div className='flex flex-col gap-y-2'>
            <label className='required-field' htmlFor='submitterEmail'>
              {format('ui.candidateDataset.submitter.hint')}
            </label>
            <Input
              {...register('submitterEmail', { required: format('validation.required') })}
              id='submitterEmail'
              placeholder={format('ui.candidateDataset.submitter.hint')}
              isInvalid={errors.submitterEmail}
            />
            {errors.submitterEmail && <ValidationError value={errors.submitterEmail?.message} />}
          </div>
          <CustomMCaptcha setCaptchaToken={setCaptchaToken} />
          <div className='flex flex-wrap text-base mt-6 gap-3'>
            <button
              type='submit'
              className='submit-button'
              disabled={mutating || reverting || !captchaToken}
            >
              {`${format('app.submit')} ${format('ui.dataset.label')}`}
              {mutating && <FaSpinner className='spinner ml-3' />}
            </button>
            {dataset &&
              <button
                type='button'
                className='cancel-button'
                disabled={mutating || reverting}
                onClick={cancelForm}
              >
                {format('app.cancel')}
                {reverting && <FaSpinner className='spinner ml-3' />}
              </button>
            }
          </div>
        </div>
      </div>
    </form>
  )
})

DatasetForm.displayName = 'DatasetForm'

export default DatasetForm
