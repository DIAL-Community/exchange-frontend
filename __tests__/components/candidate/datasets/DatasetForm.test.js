import { fireEvent, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { act } from 'react-dom/test-utils'
import { render, waitForAllEffects } from '../../../test-utils'
import { mockNextAuthUseSession, mockNextUseRouter, statuses } from '../../../utils/nextMockImplementation'
import CustomMockedProvider from '../../../utils/CustomMockedProvider'
import DatasetForm from '../../../../components/candidate/datasets/DatasetForm'

mockNextUseRouter()
describe('Unit tests for the DatasetForm component.', () => {
  const SUBMIT_BUTTON_TEST_ID = 'submit-button'
  const CANDIDATE_DATASET_NAME_TEST_ID = 'candidate-dataset-name'
  const CANDIDATE_DATASET_URL_TEST_ID = 'candidate-dataset-website'
  const CANDIDATE_DATASET_EMAIL_TEST_ID = 'candidate-dataset-email'
  const CANDIDATE_DATASET_DESCRIPTION_TEST_ID = 'candidate-dataset-description'
  const REQUIRED_FIELD_MESSAGE = 'This field is required'

  test('Should match snapshot - candidate open data.', () => {
    mockNextAuthUseSession(statuses.AUTHENTICATED)
    const { container } = render(
      <CustomMockedProvider>
        <DatasetForm />
      </CustomMockedProvider>
    )
    expect(container).toMatchSnapshot()
  })

  test('Should render Unauthorized component for unauthorized user.', async () => {
    mockNextAuthUseSession(statuses.UNAUTHENTICATED)
    const { container } = render(
      <CustomMockedProvider>
        <DatasetForm />
      </CustomMockedProvider>
    )
    await waitForAllEffects()
    expect(container).toHaveTextContent('You are not authorized to view this page')
  })

  test('Should not show validation errors for mandatory fields.', async () => {
    mockNextAuthUseSession(statuses.AUTHENTICATED)
    const user = userEvent.setup()
    const { getByTestId } = render(
      <CustomMockedProvider>
        <DatasetForm />
      </CustomMockedProvider>
    )
    await user.type(screen.getByLabelText(/Name/), 'test name')
    expect(getByTestId(CANDIDATE_DATASET_NAME_TEST_ID)).not.toHaveTextContent(REQUIRED_FIELD_MESSAGE)

    await user.type(screen.getByLabelText(/Data URL/), 'test@test.com')
    expect(getByTestId(CANDIDATE_DATASET_URL_TEST_ID)).not.toHaveTextContent(REQUIRED_FIELD_MESSAGE)

    await user.type(screen.getByLabelText(/Email Address/), 'test@test.com')
    expect(getByTestId(CANDIDATE_DATASET_EMAIL_TEST_ID)).not.toHaveTextContent(REQUIRED_FIELD_MESSAGE)

    await act(() => fireEvent.submit(getByTestId(SUBMIT_BUTTON_TEST_ID)))
    expect(getByTestId(CANDIDATE_DATASET_NAME_TEST_ID)).not.toHaveTextContent(REQUIRED_FIELD_MESSAGE)
    expect(getByTestId(CANDIDATE_DATASET_URL_TEST_ID)).not.toHaveTextContent(REQUIRED_FIELD_MESSAGE)
    expect(getByTestId(CANDIDATE_DATASET_EMAIL_TEST_ID)).not.toHaveTextContent(REQUIRED_FIELD_MESSAGE)
    expect(getByTestId(CANDIDATE_DATASET_DESCRIPTION_TEST_ID)).toHaveTextContent(REQUIRED_FIELD_MESSAGE)
  })

  test('Should show validation errors for mandatory fields and hide them on input value change.', async () => {
    mockNextAuthUseSession(statuses.AUTHENTICATED)
    const user = userEvent.setup()
    const { getByTestId } = render(
      <CustomMockedProvider>
        <DatasetForm />
      </CustomMockedProvider>
    )
    await act(() => fireEvent.submit(getByTestId(SUBMIT_BUTTON_TEST_ID)))
    expect(getByTestId(CANDIDATE_DATASET_NAME_TEST_ID)).toHaveTextContent(REQUIRED_FIELD_MESSAGE)
    expect(getByTestId(CANDIDATE_DATASET_DESCRIPTION_TEST_ID)).toHaveTextContent(REQUIRED_FIELD_MESSAGE)
    expect(getByTestId(CANDIDATE_DATASET_URL_TEST_ID)).toHaveTextContent(REQUIRED_FIELD_MESSAGE)

    await user.type(screen.getByLabelText(/Name/), 'test name')
    expect(getByTestId(CANDIDATE_DATASET_NAME_TEST_ID)).not.toHaveTextContent(REQUIRED_FIELD_MESSAGE)
    await user.clear(screen.getByLabelText(/Name/))
    expect(getByTestId(CANDIDATE_DATASET_NAME_TEST_ID)).toHaveTextContent(REQUIRED_FIELD_MESSAGE)

    await user.type(screen.getByLabelText(/Name/), 'test product name 2')
    expect(getByTestId(CANDIDATE_DATASET_NAME_TEST_ID)).not.toHaveTextContent(REQUIRED_FIELD_MESSAGE)

    await user.type(screen.getByLabelText(/Data URL/), 'test@test.com')
    expect(getByTestId(CANDIDATE_DATASET_URL_TEST_ID)).not.toHaveTextContent(REQUIRED_FIELD_MESSAGE)

    await user.type(screen.getByLabelText(/Email Address/), 'test@test.com')
    expect(getByTestId(CANDIDATE_DATASET_EMAIL_TEST_ID)).not.toHaveTextContent(REQUIRED_FIELD_MESSAGE)
    expect(getByTestId(CANDIDATE_DATASET_NAME_TEST_ID)).not.toHaveTextContent(REQUIRED_FIELD_MESSAGE)
    expect(getByTestId(CANDIDATE_DATASET_DESCRIPTION_TEST_ID)).toHaveTextContent(REQUIRED_FIELD_MESSAGE)

    await act(() => fireEvent.submit(getByTestId(SUBMIT_BUTTON_TEST_ID)))
    expect(getByTestId(CANDIDATE_DATASET_NAME_TEST_ID)).not.toHaveTextContent(REQUIRED_FIELD_MESSAGE)
    expect(getByTestId(CANDIDATE_DATASET_EMAIL_TEST_ID)).not.toHaveTextContent(REQUIRED_FIELD_MESSAGE)
    expect(getByTestId(CANDIDATE_DATASET_DESCRIPTION_TEST_ID)).toHaveTextContent(REQUIRED_FIELD_MESSAGE)
  })
})
