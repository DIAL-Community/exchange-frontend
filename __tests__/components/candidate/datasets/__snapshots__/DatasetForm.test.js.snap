// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Unit tests for the DatasetForm component. Should match snapshot - candidate open data. 1`] = `
<div>
  <div
    class="Toastify"
  />
  <div
    class="flex flex-col"
  >
    <div
      class="hidden lg:block px-8"
    >
      <div
        class="bg-white pb-3 lg:py-4 whitespace-nowrap text-ellipsis overflow-hidden"
      >
        <a
          class="inline text-dial-blue h5"
          href="/"
        >
          Home
        </a>
        <div
          class="inline h5"
        >
           &gt; 
          <a
            class="text-dial-blue"
            href="/datasets"
          >
            Open Data
          </a>
           &gt; 
          <span
            class="text-dial-gray-dark"
          >
            Create
          </span>
        </div>
      </div>
    </div>
    <div
      class="pb-8 px-8"
    >
      <div
        class="sm:px-0 max-w-full mx-auto"
        id="content"
      >
        <form>
          <div
            class="bg-edit shadow-md rounded px-8 pt-6 pb-12 mb-4 flex flex-col gap-3"
          >
            <div
              class="text-2xl font-bold text-dial-blue pb-4"
            >
              Candidate Open Data
            </div>
            <div
              class="flex flex-col lg:flex-row gap-4"
            >
              <div
                class="w-full lg:w-1/2 flex flex-col gap-y-3"
              >
                <div
                  class="form-field-wrapper"
                  data-testid="candidate-dataset-name"
                >
                  <label
                    class="form-field-label required-field"
                    for="name"
                  >
                    Dataset Name
                  </label>
                  <input
                    class="w-full"
                    data-testid="dataset-name-input"
                    id="name"
                    name="name"
                    placeholder="Dataset Name"
                    type="text"
                    value=""
                  />
                </div>
                <div
                  class="form-field-wrapper"
                  data-testid="candidate-dataset-dataUrl"
                >
                  <label
                    class="form-field-label required-field"
                    for="dataUrl"
                  >
                    Data URL
                  </label>
                  <div
                    class="border border-transparent url-input bg-white flex items-center"
                  >
                    <input
                      class="url w-full placeholder-dial-gray focus:outline-0"
                      id="dataUrl"
                      placeholder="Data URL"
                      value=""
                    />
                  </div>
                </div>
                <div
                  class="form-field-wrapper"
                >
                  <label
                    class="form-field-label"
                    for="dataVisualizationUrl"
                  >
                    Dataset Visualization URL
                  </label>
                  <div
                    class="border border-transparent url-input bg-white flex items-center"
                  >
                    <input
                      class="url w-full placeholder-dial-gray focus:outline-0"
                      id="dataVisualizationUrl"
                      placeholder="Dataset Visualization URL"
                      value=""
                    />
                  </div>
                </div>
                <div
                  class="form-field-wrapper"
                >
                  <label
                    class="form-field-label"
                  >
                    Dataset Type
                  </label>
                  <div
                    class=" css-l45kjq-container"
                  >
                    <span
                      class="css-1f43avz-a11yText-A11yText"
                      id="react-select-:r0:-live-region"
                    />
                    <span
                      aria-atomic="false"
                      aria-live="polite"
                      aria-relevant="additions text"
                      class="css-1f43avz-a11yText-A11yText"
                    />
                    <div
                      class="react-select__control css-19jej77-control"
                    >
                      <div
                        class="react-select__value-container react-select__value-container--has-value css-3tzzl5-ValueContainer"
                      >
                        <div
                          class="react-select__single-value css-qc6sy-singleValue"
                        >
                          Dataset
                        </div>
                        <div
                          class="react-select__input-container css-6j8wv5-Input"
                          data-value=""
                        >
                          <input
                            aria-autocomplete="list"
                            aria-expanded="false"
                            aria-haspopup="true"
                            autocapitalize="none"
                            autocomplete="off"
                            autocorrect="off"
                            class="react-select__input"
                            id="react-select-id-dataType"
                            role="combobox"
                            spellcheck="false"
                            style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
                            tabindex="0"
                            type="text"
                            value=""
                          />
                        </div>
                      </div>
                      <div
                        class="react-select__indicators css-1hb7zxy-IndicatorsContainer"
                      >
                        <span
                          class="react-select__indicator-separator css-1979yuk-indicatorSeparator"
                        />
                        <div
                          aria-hidden="true"
                          class="react-select__indicator react-select__dropdown-indicator css-1hmmjq6-indicatorContainer"
                        >
                          <svg
                            aria-hidden="true"
                            class="css-tj5bde-Svg"
                            focusable="false"
                            height="20"
                            viewBox="0 0 20 20"
                            width="20"
                          >
                            <path
                              d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                            />
                          </svg>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
                <div
                  class="form-field-wrapper"
                  data-testid="candidate-dataset-email"
                >
                  <label
                    class="form-field-label required-field"
                    for="email"
                  >
                    Email Address
                  </label>
                  <input
                    class="w-full"
                    id="email"
                    name="submitterEmail"
                    placeholder="Email Address"
                    type="text"
                    value=""
                  />
                </div>
                <div
                  class="form-field-wrapper"
                >
                  <div />
                </div>
              </div>
              <div
                class="w-full lg:w-1/2"
              >
                <div
                  class="form-field-wrapper"
                  data-testid="candidate-dataset-description"
                >
                  <label
                    class="form-field-label required-field"
                  >
                    Description
                  </label>
                  <div
                    class="relative"
                  >
                    <svg
                      class="absolute text-lg inset-x-1/2 top-10 spinner"
                      fill="currentColor"
                      height="2em"
                      stroke="currentColor"
                      stroke-width="0"
                      viewBox="0 0 512 512"
                      width="2em"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M304 48c0 26.51-21.49 48-48 48s-48-21.49-48-48 21.49-48 48-48 48 21.49 48 48zm-48 368c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48zm208-208c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48zM96 256c0-26.51-21.49-48-48-48S0 229.49 0 256s21.49 48 48 48 48-21.49 48-48zm12.922 99.078c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48c0-26.509-21.491-48-48-48zm294.156 0c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48c0-26.509-21.49-48-48-48zM108.922 60.922c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.491-48-48-48z"
                      />
                    </svg>
                    <div
                      class="html-editor"
                    >
                      <textarea
                        id="description-editor"
                        style="visibility: hidden;"
                      />
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="flex flex-wrap text-xl mt-8 gap-3"
            >
              <button
                class="submit-button"
                data-testid="submit-button"
                type="submit"
              >
                Submit Candidate Open Data
              </button>
              <button
                class="cancel-button"
                type="button"
              >
                Cancel
              </button>
            </div>
          </div>
        </form>
      </div>
    </div>
  </div>
</div>
`;
